### ISSUE TITLE
Deny when reference to result of unsafe stored somewhere

### ISSUE THREAD

**Issue by AngelicosPhosphoros** [rust-lang/rust-clippy#7311]
### What it does

It should deny this code:

```rust
struct MyStruct{
   field: usize
}

let my_struct_ptr: *mut MyStruct = ...;
let reference_to_field = &mut unsafe { (*my_struct_ptr).field };
// use reference here
```
This code actually creates temporary with value of field, takes reference to it, then drops temporary.
Therefore here we have dangling pointer to stack.

It should suggest this:
```
let reference_to_field = unsafe { &mut (*my_struct_ptr).field };
```

### Categories (optional)

- Kind: clippy::correctness
- Remove UB code created in unsafe block

### Drawbacks

None.

### Example

```rust
let reference_to_field = &mut unsafe { (*my_struct_ptr).field };
```

Should be written as:

```rust
let reference_to_field = unsafe { &mut (*my_struct_ptr).field };
```

### Also

Link to issue which caused by original code: https://github.com/Morganamilo/paru/issues/392
Bugfix: https://github.com/archlinux/alpm.rs/commit/5253d6
Relevant [Reddit discussion](https://www.reddit.com/r/rust/comments/nqjyb7/the_most_annoying_bug_ive_had_to_track_down/).

**Comment by AngelicosPhosphoros** [comment]
Well, actually it isn't caused by unsafe but by implicit conversion from reference to pointer.

So I agree with this [comment](https://www.reddit.com/r/rust/comments/nqjyb7/the_most_annoying_bug_ive_had_to_track_down/h0brwp8?utm_source=share&utm_medium=web2x&context=3):

> I think the lint should just forbid storing a reference to a temporary in a pointer, since it's error-prone. If you have a valid use case, explicitly put a temporary in a local variable. It will also be more readable.

Example (from this [comment](https://www.reddit.com/r/rust/comments/nqjyb7/the_most_annoying_bug_ive_had_to_track_down/h0b5jyh?utm_source=share&utm_medium=web2x&context=3)):
```rust
struct A<'a> {
    val: &'a mut u8
}

struct B {
    val: *mut u8
}

fn build_a<'a>() -> A<'a> {
    // Here borrow checker prevents it
    A { val: &mut 0 }
}

fn build_b() -> B {
    // Here it is not because reference converts to pointer
    B { val: &mut 0 }
}

fn main() {
}
```

**Comment by CAD97** [comment]
I think it's important to note that `let reference_to_field = &mut unsafe { (*my_struct_ptr).field };` _does not_ create a dangling reference; the temporary is kept alive until the reference is dead via temporary lifetime extension. The issue only comes when converting the reference to a pointer that outlives the reference.

As such I agree that the lint should only trigger when a reference to a temporary is converted to a pointer. It could potentially also be restricted to a simple escape test; if the pointer is only used as a local and never used as a function argument, it could not be misused (at least per my understanding of when an extended temporary is dropped - as any other local, at the end of scope).

(A restriction lint could also be created for taking a reference to a temporary beyond the simple use as a function argument, as such references are easy to misuse and accidentally change temporary drop ordering in general.)

**Comment by Cryptjar** [comment]
I also agree that this is only an issue with pointers (i.e. `let reference_to_field: *mut _ = &mut ...`). In this context, I would like to point to the [`std::ptr::addr_of`](https://doc.rust-lang.org/std/ptr/macro.addr_of.html) and [`std::ptr::addr_of_mut`](https://doc.rust-lang.org/std/ptr/macro.addr_of_mut.html) macros, which in my opinion should be _always_ preferred over casting ad-hoc references, i.e. `addr_of_mut!(...)`  instead of `(&mut ...) as *mut _` .

On that notice, it seems to me, that the `addr_of` macors directly deny pointers to temporaries (also see [E0745](https://doc.rust-lang.org/stable/error-index.html#E0745)). That means that `let reference_to_field: *mut _ = addr_of_mut!(unsafe { (*my_struct_ptr).field })` doesn't even compile: [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=1d79a30579766a00001b800c8f0f36ed).

So, maybe it would be better to create a lint for promoting `addr_of` over ad-hoc reference to pointer casting.
